### PR TITLE
Use MTLCompiler directly (v2)

### PR DESCRIPTION
**Pull request by uuuvn**

### CODE REVIEW

File: tinygrad/runtime/ops_metal.py [hunk 1]
```diff
@@ -65,20 +67,35 @@ def metal_src_to_library(device:MetalDevice, src:str) -> objc_instance:
...
+    params = f'-fno-fast-math -std=metal3.1 --driver-mode=metal -x metal -fmodules-cache-path="{os.path.join(_cache_dir, "tinygrad")}"'
+    # source blob has to be padded to multiple of 4 but at least one 'b\x00' should be added
+    request = struct.pack('<Q', round_up(len(src)+1, 4)) + struct.pack('<Q', len(params)+1)
+    request += src.encode() + b'\x00'*(4-len(src)%4) + params.encode() + b'\x00'
+    # 13 is requestType that metal uses for this stage of compilation, there aren't any docs or symbols.
+    # The callback is actully not a callback but a block which is apple's non-standart extension to add closures to C.
+    # See https://clang.llvm.org/docs/Block-ABI-Apple.html#high-level for struct layout.
+    # Fields other than invoke are unused in this case so we can just use ctypes.byref with negative offset to invoke field, add blockptr as a first
+    # argument and pretend it's a normal callback
+    compiler.MTLCodeGenServiceBuildRequest(ctypes.c_void_p(self.cgs), None, 13, request, len(request), ctypes.byref(callback, -0x10))
```
Comment (nimlgen) [inline tinygrad/runtime/ops_metal.py:95]: COMPILE_REQUEST_TYPE=13 and comment!

the same for "closures", there will be way less questions if you write something like `_block_literal(callback)` and put all comments about layout details in `def _block_literal`

Comment (uuuvn) [reply tinygrad/runtime/ops_metal.py]: fixed

File: tinygrad/runtime/ops_metal.py [hunk 2]
```diff
@@ -65,20 +75,30 @@ def metal_src_to_library(device:MetalDevice, src:str) -> objc_instance:
...
+      if error == 0:
+        reply = bytes(to_mv(dataPtr, dataLen))
+        # offset from beginning to data = header size + warning size
+        ret = reply[sum(struct.unpack('<LL', reply[8:16])):]
+      else:
+        ret = CompileError(errorMessage.decode())
+    # llvm will create modules.timestamp in cache path and cache compilation of metal stdlib (250ms => 8ms compilation time)
+    params = f'-fno-fast-math -std=metal3.1 --driver-mode=metal -x metal -fmodules-cache-path="{os.path.join(_cache_dir, "tinygrad")}"'
+    # source blob has to be padded to multiple of 4 but at least one 'b\x00' should be added
+    request = struct.pack('<Q', round_up(len(src)+1, 4)) + struct.pack('<Q', len(params)+1)
```
Comment (geohot) [inline tinygrad/runtime/ops_metal.py:96]: Why two struct.pack?

File: tinygrad/runtime/ops_metal.py [hunk 3]
```diff
@@ -65,20 +75,30 @@ def metal_src_to_library(device:MetalDevice, src:str) -> objc_instance:
...
+    request += src.encode() + b'\x00'*(4-len(src)%4) + params.encode() + b'\x00'
```
Comment (geohot) [inline tinygrad/runtime/ops_metal.py:97]: This is computed twice

File: tinygrad/runtime/ops_metal.py [hunk 4]
```diff
@@ -65,20 +75,30 @@ def metal_src_to_library(device:MetalDevice, src:str) -> objc_instance:
...
+    ret: Union[Exception, bytes] = CompileError("MTLCodeGenServiceBuildRequest returned without calling the callback")
+    @ctypes.CFUNCTYPE(None, ctypes.c_void_p, ctypes.c_int32, ctypes.c_void_p, ctypes.c_size_t, ctypes.c_char_p)
+    def callback(blockptr, error, dataPtr, dataLen, errorMessage):
+      nonlocal ret
+      if error == 0:
+        reply = bytes(to_mv(dataPtr, dataLen))
+        # offset from beginning to data = header size + warning size
+        ret = reply[sum(struct.unpack('<LL', reply[8:16])):]
+      else:
+        ret = CompileError(errorMessage.decode())
```
Comment (geohot) [inline tinygrad/runtime/ops_metal.py:92]: Can you raise in the callback?

File: tinygrad/runtime/ops_metal.py [hunk 5]
```diff
@@ -65,20 +75,30 @@ def metal_src_to_library(device:MetalDevice, src:str) -> objc_instance:
...
+    # llvm will create modules.timestamp in cache path and cache compilation of metal stdlib (250ms => 8ms compilation time)
+    params = f'-fno-fast-math -std=metal3.1 --driver-mode=metal -x metal -fmodules-cache-path="{os.path.join(_cache_dir, "tinygrad")}"'
+    # source blob has to be padded to multiple of 4 but at least one 'b\x00' should be added, params blob just has to be null terminated
+    src_padded, params_padded = src.encode() + b'\x00'*(4-len(src)%4), params.encode() + b'\x00'
+    request = struct.pack('<2Q', len(src_padded), len(params_padded)) + src_padded + params_padded
```
Comment (geohot) [inline tinygrad/runtime/ops_metal.py:97]: is `<2Q` the same as `<QQ`?

File: tinygrad/runtime/ops_metal.py [hunk 6]
```diff
@@ -65,20 +75,30 @@ def metal_src_to_library(device:MetalDevice, src:str) -> objc_instance:
...
+      if error == 0:
+        reply = bytes(to_mv(dataPtr, dataLen))
+        # offset from beginning to data = header size + warning size
+        ret = reply[sum(struct.unpack('<LL', reply[8:16])):]
+      else:
+        ret = CompileError(errorMessage.decode())
+    # llvm will create modules.timestamp in cache path and cache compilation of metal stdlib (250ms => 8ms compilation time)
+    params = f'-fno-fast-math -std=metal3.1 --driver-mode=metal -x metal -fmodules-cache-path="{os.path.join(_cache_dir, "tinygrad")}"'
+    # source blob has to be padded to multiple of 4 but at least one 'b\x00' should be added, params blob just has to be null terminated
+    src_padded, params_padded = src.encode() + b'\x00'*(4-len(src)%4), params.encode() + b'\x00'
```
Comment (geohot) [inline tinygrad/runtime/ops_metal.py:96]: `round_up(len(src), 4)-len(src)`

File: tinygrad/runtime/ops_metal.py (hunk 5 above)
Comment (uuuvn) [reply tinygrad/runtime/ops_metal.py:97]: yes

File: tinygrad/runtime/ops_metal.py (hunk 6 above)
Comment (uuuvn) [reply tinygrad/runtime/ops_metal.py:96]: This won't null terminate if len(src) is a multiple of 4
```
>>> from tinygrad.helpers import round_up
>>> src = 'a'*16
>>> 4-len(src)%4
4
>>> round_up(len(src), 4)-len(src)
0
>>> round_up(len(src) + 1, 4)-len(src)
4
```
I can change it to `round_up(len(src) + 1, 4)-len(src)` (+1 is important) but this feels significantly less readable to me than `4-len(src)%4`

File: tinygrad/runtime/ops_metal.py [hunk 7]
```diff
@@ -45,6 +50,11 @@ class Struct(ctypes.Structure): pass
   Struct._fields_ = [(f"field{i}", _type) for i in range(len(t))]
   return Struct(*t)
 
+def to_block_literal(callback):
```
[the remark this answers is not in the document]
Comment (uuuvn) [reply tinygrad/runtime/ops_metal.py:53]: > the same for "closures", there will be way less questions if you write something like `_block_literal(callback)` and put all comments about layout details in `def _block_literal`

https://github.com/tinygrad/tinygrad/pull/7920#discussion_r1860954360

Should i revert this change?

**Comment by geohot** [comment]
Wow this actually looks a lot faster

master:
IGNORE_BEAM_CACHE=1 DEBUG=2 BEAM=2 python3 examples/efficientnet.py  201.02s user 16.31s system 144% cpu 2:30.44 total

this:
IGNORE_BEAM_CACHE=1 DEBUG=2 BEAM=2 python3 examples/efficientnet.py  214.95s user 18.58s system 364% cpu 1:04.03 total

File: tinygrad/runtime/ops_metal.py [hunk 8]
```diff
@@ -65,20 +70,34 @@ def metal_src_to_library(device:MetalDevice, src:str) -> objc_instance:
...
+    def callback(blockptr, error, dataPtr, dataLen, errorMessage):
+      nonlocal ret
+      if error == 0:
+        reply = bytes(to_mv(dataPtr, dataLen))
+        # offset from beginning to data = header size + warning size
+        ret = reply[sum(struct.unpack('<LL', reply[8:16])):]
+      else:
+        ret = CompileError(errorMessage.decode())
+    # llvm will create modules.timestamp in cache path and cache compilation of metal stdlib (250ms => 8ms compilation time)
+    params = f'-fno-fast-math -std=metal3.1 --driver-mode=metal -x metal -fmodules-cache-path="{os.path.join(_cache_dir, "tinygrad")}"'
```
[the remark this answers is not in the document]
Comment (uuuvn) [reply tinygrad/runtime/ops_metal.py:89]: It is 0 bytes because it is modules.**timestamp**
It doesn't create cache for stdlib because it is precompiled. On my system it is in `/System/Library/PrivateFrameworks/GPUCompiler.framework/Versions/32023/Libraries/lib/clang/32023.404/include/metal/prebuilt_implicit_modules/3EN3BEEDJJAKK/monolithic_metal.pcm`
We have to specify modules cache anyways path because otherwise it doesn't use modules cache
You can verify that by adding `-fno-prebuilt-implicit-modules` to `params` and it will create `Library/Caches/tinygrad/3EN3BEEDJJAKK/{modules.idx,monolithic_metal.pcm}` (106k and 7.4M)

**Comment by geohot** [comment]
You still have a typo. Again, this back and forth is very frustrating.

I copy and pasted the code into chatgpt and it found it. https://chatgpt.com/share/67500421-22e0-8000-a7b6-d91c1dca7b44

Please stop wasting my time with something you haven't sanity checked in an LLM.

File: tinygrad/runtime/ops_metal.py [hunk 9]
```diff
@@ -65,20 +70,35 @@ def metal_src_to_library(device:MetalDevice, src:str) -> objc_instance:
...
+        ret = CompileError(errorMessage.decode())
+    # llvm will create modules.timestamp in cache path and cache compilation of metal stdlib (250ms => 8ms compilation time)
+    # note that llvm won't necessarily create anything else here as apple has prebuilt versions of many standart libraries
+    params = f'-fno-fast-math -std=metal3.1 --driver-mode=metal -x metal -fmodules-cache-path="{os.path.join(_cache_dir, "tinygrad")}"'
+    # source blob has to be padded to multiple of 4 but at least one 'b\x00' should be added, params blob just has to be null terminated
+    src_padded, params_padded = src.encode() + b'\x00'*(round_up(len(src) + 1, 4) - len(src)), params.encode() + b'\x00'
+    request = struct.pack('<QQ', len(src_padded), len(params_padded)) + src_padded + params_padded
+    # The callback is actually not a callback but a block which is apple's non-standart extension to add closures to C.
+    # See https://clang.llvm.org/docs/Block-ABI-Apple.html#high-level for struct layout.
+    # Fields other than invoke are unused in this case so we can just use ctypes.byref with negative offset to invoke field, add blockptr as a first
```
Comment (geohot) [inline tinygrad/runtime/ops_metal.py:96]: ugh make the struct, don't do this.

File: tinygrad/runtime/ops_metal.py [hunk 10]
```diff
@@ -176,7 +196,7 @@ def __init__(self, device:str):
     self.timeline_value = 0
 
     from tinygrad.runtime.graph.metal import MetalGraph
-    super().__init__(device, MetalAllocator(self), MetalRenderer(), MetalCompiler() if getenv("METAL_XCODE") else Compiler(),
+    super().__init__(device, MetalAllocator(self), MetalRenderer(), MetalCompiler() if getenv("METAL_DIRECT", 1) else Compiler(), \
```
Comment (geohot) [inline tinygrad/runtime/ops_metal.py:199]: It's little super frustrating things. Why did you add a \ here? Like read the diff over and over.

**Comment by geohot** [comment]
Okay, I will merge when CI passes and give you the bounty. I want this for the search speed. Can pay via PayPal or USDC on ETH.

However, if I see any more PRs that haven't been carefully read before they are submitted, they will immediately not be bounty eligible.
